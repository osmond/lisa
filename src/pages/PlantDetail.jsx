import { useParams, Link } from 'react-router-dom'
import { useState, useRef } from 'react'
import { usePlants } from '../PlantContext.jsx'

export default function PlantDetail() {
  const { id } = useParams()
  const { plants, addPhoto, removePhoto } = usePlants()
  const plant = plants.find(p => p.id === Number(id))

  const tabNames = ['activity', 'notes', 'care']
  const tabRefs = useRef([])
  const [tab, setTab] = useState('activity')
  const fileInputRef = useRef()

  const handleFiles = e => {
    const files = Array.from(e.target.files || [])
    files.forEach(file => {
      const reader = new FileReader()
      reader.onload = ev => addPhoto(plant.id, ev.target.result)
      reader.readAsDataURL(file)
    })
    e.target.value = ''
  }

  const handleKeyDown = (e, index) => {
    if (e.key === 'ArrowRight' || e.key === 'ArrowLeft') {
      e.preventDefault()
      const dir = e.key === 'ArrowRight' ? 1 : -1
      const nextIndex = (index + dir + tabNames.length) % tabNames.length
      setTab(tabNames[nextIndex])
      tabRefs.current[nextIndex]?.focus()
    }
  }

  if (!plant) {
    return <div className="text-gray-700">Plant not found</div>
  }

  return (
    <div className="space-y-2">
      <div className="space-y-4">
        <img src={plant.image} alt={plant.name} loading="lazy" className="w-full h-64 object-cover rounded-xl" />
        <div>
          <h1 className="text-2xl font-bold">{plant.name}</h1>
          {plant.nickname && <p className="text-gray-500">{plant.nickname}</p>}
        </div>

        <div className="grid gap-1 text-sm">
          <p><strong>Last watered:</strong> {plant.lastWatered}</p>
          <p><strong>Next watering:</strong> {plant.nextWater}</p>
          {plant.lastFertilized && (
            <p><strong>Last fertilized:</strong> {plant.lastFertilized}</p>
          )}
        </div>

        <div className="grid gap-1 text-sm">
          {plant.light && <p><strong>Light:</strong> {plant.light}</p>}
          {plant.humidity && <p><strong>Humidity:</strong> {plant.humidity}</p>}
          {plant.difficulty && <p><strong>Difficulty:</strong> {plant.difficulty}</p>}
        </div>

        <div>
          <Link
            to={`/plant/${plant.id}/gallery`}
            className="text-green-600 underline"
          >
            View Gallery
          </Link>
        </div>

        <div>
          <div className="flex space-x-4 border-b" role="tablist">
            <button
              role="tab"
              ref={el => (tabRefs.current[0] = el)}
              aria-selected={tab === 'activity'}
              tabIndex={tab === 'activity' ? 0 : -1}
              className={`py-2 ${tab === 'activity' ? 'border-b-2 border-green-500 font-medium' : ''}`}
              onClick={() => setTab('activity')}
              onKeyDown={e => handleKeyDown(e, 0)}
            >
              Activity
            </button>
            <button
              role="tab"
              ref={el => (tabRefs.current[1] = el)}
              aria-selected={tab === 'notes'}
              tabIndex={tab === 'notes' ? 0 : -1}
              className={`py-2 ${tab === 'notes' ? 'border-b-2 border-green-500 font-medium' : ''}`}
              onClick={() => setTab('notes')}
              onKeyDown={e => handleKeyDown(e, 1)}
            >
              Notes
            </button>
            <button
              role="tab"
              ref={el => (tabRefs.current[2] = el)}
              aria-selected={tab === 'care'}
              tabIndex={tab === 'care' ? 0 : -1}
              className={`py-2 ${tab === 'care' ? 'border-b-2 border-green-500 font-medium' : ''}`}
              onClick={() => setTab('care')}
              onKeyDown={e => handleKeyDown(e, 2)}
            >
              Advanced
            </button>
          </div>
          <div className="p-4">
            {tab === 'activity' && (
              <ul className="list-disc pl-4 space-y-1">
                {(plant.activity || []).map((a, i) => (
                  <li key={i}>{a}</li>
                ))}
              </ul>
            )}
            {tab === 'notes' && (
              <div>{plant.notes || 'No notes yet.'}</div>
            )}
            {tab === 'care' && (
              <div>{plant.advancedCare || 'No advanced care info.'}</div>
            )}
        </div>
      </div>
      <div className="space-y-2">
        <h2 className="text-lg font-semibold">Gallery</h2>
        <div className="grid grid-cols-3 gap-2">
          {(plant.gallery || []).map((src, i) => (
            <div key={i} className="relative">
              <img
                src={src}
                alt={`${plant.name} ${i}`}
                className="object-cover w-full h-24 rounded"
              />
              <button
                className="absolute top-1 right-1 bg-white bg-opacity-70 rounded px-1 text-xs"
                onClick={() => removePhoto(plant.id, i)}
              >
                âœ•
              </button>
            </div>
          ))}
        </div>
        <button
          type="button"
          onClick={() => fileInputRef.current.click()}
          className="mt-2 px-3 py-1 bg-green-600 text-white rounded"
        >
          Add Photo
        </button>
        <input
          type="file"
          accept="image/*"
          multiple
          ref={fileInputRef}
          onChange={handleFiles}
          className="hidden"
        />
      </div>
    </div>
  </div>
)
}
