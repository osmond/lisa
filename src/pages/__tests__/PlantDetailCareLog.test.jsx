import { render, screen } from '@testing-library/react'
import { MemoryRouter, Routes, Route } from 'react-router-dom'
import PlantDetail from '../PlantDetail.jsx'

let mockPlants = []
jest.mock('../../PlantContext.jsx', () => ({
  usePlants: () => ({
    plants: mockPlants,
    addPhoto: jest.fn(),
    removePhoto: jest.fn(),
  }),
}))

beforeEach(() => {
  mockPlants = [
    {
      id: 1,
      name: 'Plant A',
      image: 'a.jpg',
      careLog: [
        { date: '2025-07-02', type: 'Watered', note: 'deep soak' },
      ],
      photos: [],
    },
  ]
})

test('shows notes from care log in activity tab', () => {
  render(
    <MemoryRouter initialEntries={['/plant/1']}>
      <Routes>
        <Route path="/plant/:id" element={<PlantDetail />} />
      </Routes>
    </MemoryRouter>
  )

  expect(screen.getByText('Watered on 2025-07-02 - deep soak')).toBeInTheDocument()
})
