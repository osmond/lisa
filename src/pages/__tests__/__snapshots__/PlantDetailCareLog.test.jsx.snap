// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`timeline bullet markup matches snapshot 1`] = `
<ul
  class=" relative ml-3 space-y-8 pl-5 before:absolute before:inset-y-0 before:left-2 before:w-px before:bg-gray-200"
>
  <li
    class="relative text-xs sm:text-sm"
  >
    <div
      class="absolute -left-5 top-[0.25rem] w-4 h-4 flex items-center justify-center rounded-full bg-green-500 z-10"
    >
      <svg
        aria-hidden="true"
        class="w-3 h-3 text-white"
        fill="currentColor"
        height="1em"
        viewBox="0 0 256 256"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <rect
          fill="none"
          height="256"
          width="256"
        />
        <line
          fill="none"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="16"
          x1="112"
          x2="176"
          y1="112"
          y2="112"
        />
        <line
          fill="none"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="16"
          x1="112"
          x2="176"
          y1="144"
          y2="144"
        />
        <rect
          fill="none"
          height="176"
          rx="8"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="16"
          width="176"
          x="40"
          y="40"
        />
        <line
          fill="none"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="16"
          x1="80"
          x2="80"
          y1="40"
          y2="216"
        />
      </svg>
    </div>
    <div
      class="flex items-start bg-gray-50 dark:bg-gray-700 rounded-xl p-3 shadow-sm"
    >
      <div>
        <span
          class="font-medium"
        >
          July 2, 2025
        </span>
         â€” 
        Watered
        <div
          class="text-xs italic text-green-700 mt-1"
        >
          deep soak
        </div>
      </div>
    </div>
  </li>
</ul>
`;
